File: composeApp/src/iosMain/kotlin/dev/tcode/thinmpk/repository/SongRepository.kt
package dev.tcode.thinmpk.repository

import dev.tcode.thinmpk.model.SongModel
import platform.MediaPlayer.MPMediaItem
import platform.MediaPlayer.MPMediaItemCollection
import platform.MediaPlayer.MPMediaItemPropertyIsCloudItem
import platform.MediaPlayer.MPMediaPropertyPredicate
import platform.MediaPlayer.MPMediaQuery

actual class SongRepository actual constructor() {
    fun findMediaItemById(id: String): MPMediaItem? {
        val query = MPMediaQuery.songsQuery()
        query.addFilterPredicate(
            MPMediaPropertyPredicate.predicateWithValue(
                value = false,
                forProperty = MPMediaItemPropertyIsCloudItem
            )
        )

        val collections = query.collections ?: return null
        val collection = collections.firstOrNull { c ->
            val mediaCollection = c as? MPMediaItemCollection ?: return@firstOrNull false
            id == mediaCollection.representativeItem?.persistentID?.toString()
        } as? MPMediaItemCollection ?: return null

        return collection.representativeItem
    }

    actual fun findById(id: String): SongModel? {
        val mediaItem = findMediaItemById(id) ?: return null

        return SongModel(
            id = mediaItem.persistentID.toString(),
            name = mediaItem.title ?: "",
            artistName = mediaItem.artist ?: "",
            albumId = mediaItem.albumPersistentID.toString(),
            albumName = mediaItem.albumTitle ?: "",
            imageId = mediaItem.persistentID.toString(),
            duration = mediaItem.playbackDuration.toInt()
        )
    }

    actual fun findAll(): List<SongModel> {
        val query = MPMediaQuery.songsQuery()
        val predicate = MPMediaPropertyPredicate.predicateWithValue(
            value = false,
            forProperty = MPMediaItemPropertyIsCloudItem
        )
        query.addFilterPredicate(predicate)

        val collections = query.collections ?: return emptyList()

        return collections.mapNotNull { collection ->
            val mediaCollection = collection as? MPMediaItemCollection ?: return@mapNotNull null
            val mediaItem = mediaCollection.representativeItem ?: return@mapNotNull null
            SongModel(
                id = mediaItem.persistentID.toString(),
                name = mediaItem.title ?: "",
                artistName = mediaItem.artist ?: "",
                albumId = mediaItem.albumPersistentID.toString(),
                albumName = mediaItem.albumTitle ?: "",
                imageId = mediaItem.persistentID.toString(),
                duration = mediaItem.playbackDuration.toInt()
            )
        }
    }
}

File: iosApp/iosApp/SongRepository.swift
import MediaPlayer
import ComposeApp

class SongRepository {
    func findAll() -> [SongModel] {
        let query = MPMediaQuery.songs()
        query.addFilterPredicate(
            MPMediaPropertyPredicate(
                value: false,
                forProperty: MPMediaItemPropertyIsCloudItem
            )
        )

        guard let collections = query.collections else { return [] }

        return collections.compactMap { collection in
            guard let mediaItem = collection.representativeItem else { return nil }
            return SongModel(
                id: String(mediaItem.persistentID),
                name: mediaItem.title ?? "",
                artistName: mediaItem.artist ?? "",
                albumId: String(mediaItem.albumPersistentID),
                albumName: mediaItem.albumTitle ?? "",
                duration: Int32(mediaItem.playbackDuration),
                imageId: String(mediaItem.persistentID)
            )
        }
    }

    func findById(id: String) -> SongModel? {
        let query = MPMediaQuery.songs()
        query.addFilterPredicate(
            MPMediaPropertyPredicate(
                value: false,
                forProperty: MPMediaItemPropertyIsCloudItem
            )
        )

        guard let collections = query.collections else { return nil }

        guard let collection = collections.first(where: {
            $0.representativeItem?.persistentID.description == id
        }) else { return nil }

        guard let mediaItem = collection.representativeItem else { return nil }

        return SongModel(
            id: String(mediaItem.persistentID),
            name: mediaItem.title ?? "",
            artistName: mediaItem.artist ?? "",
            albumId: String(mediaItem.albumPersistentID),
            albumName: mediaItem.albumTitle ?? "",
            duration: Int32(mediaItem.playbackDuration),
            imageId: String(mediaItem.persistentID)
        )
    }
}
